### ISSUE TITLE
Could reportUnboundVariable take code flow into account?

### ISSUE THREAD

**Issue by bobbyrenwick** [microsoft/pyright#1540]
**Is your feature request related to a problem? Please describe.**

```py
from typing import Union
from typing_extensions import Literal

def test_out_unions(my_var: Union[Literal["one"], Literal["two"]]):
    if my_var == "one":
        other_var = 1
    elif my_var == "two":
        other_var = 2

    return other_var <----------- other_var is possibly unbound
```

**Describe the solution you'd like**
Given that you've exhausted all the possible types for `my_var`, you know that `other_var` will be bound. I think there's something similar in TS?

**Comment by erictraut** [comment]
This check does take code flow into account, but the code flow graph is not affected by type narrowing (in this case of `my_var`). The code flow graph is an input to the type analyzer, and it must be constructed prior to type analysis. 

I'll think about this more to see if I can come up with a way around this chicken-and-egg problem, but I'm not optimistic. Pyright's architecture is based on the one used in the TypeScript compiler, and it has this same limitation.

**Comment by erictraut** [comment]
Here is the same function in TypeScript.

```typescript
export function foo(a: 'one' | 'two') {
    let b: number;

    if (a === 'one') {
        b = 1;
    } else if (a === 'two') {
        b = 2;
    }

    return b; // Error: Variable 'b' is used before being assigned
}
```

**Comment by erictraut** [comment]
Here's an approach that is recommended in mypy forums.

```python
from typing import Literal, NoReturn, Union


def assert_never(val: NoReturn) -> NoReturn:
    """
    Use this utility function to statically (and dynamically)
    verify that the specified value has been narrowed to
    a "Never" type.
    """
    raise Exception("We should never get here")

def test_out_unions(my_var: Union[Literal["one"], Literal["two"]]):
    if my_var == "one":
        other_var = 1
    elif my_var == "two":
        other_var = 2
    else:
        assert_never(my_var)

    return other_var
```

**Comment by bobbyrenwick** [comment]
@erictraut how right you are that TS does the same thing. I guess I end up writing fewer of these cases in TS because you have to declare the variable and therefore its type at the top of the block. Whereas in python you can't do that! I've just activated `pyright` on the codebase (coming from mypy and it's 2x faster, thanks 👍🏻 🚀 ) and there were a few possible unbound errrors, all with similar logic of exhausting the possible types.

The mypy forum suggestion works in removing the possible unbound error. It also works in TS if you throw an error in the else case. What isn't immediately clear to to me is how that error raising then propagates the fact that `other_var` will be set in the other cases, but maybe I just need to change my mental model to allow for that.

Is there any way to import the Never type so I can have the type system throw an error in the `else` case, as well as a runtime error, so that if you were to add `Literal["three"]` as possibility it'd cause an error as `my_var` would have type `Literal["three"]` and being assigned to `assert_never`'s `arg: Never`?

Here's a similar TS example - https://stackoverflow.com/a/39419171 of the `never` arg type.

**Comment by erictraut** [comment]
>What isn't immediately clear to to me is how that error raising then propagates the fact that other_var will be set in the other cases.

The reason this works is that it modifies the code flow graph to disallow a "fall through". All of the paths in the code flow graph that lead to the return statement set the value of `other_var`.

`Never` isn't a type that is supported by Python at runtime. It's also not documented in PEP 484 or any other Python specs. It's a concept I borrowed from TypeScript, and it's used only internally within Pyright. It's semantically equivalent to an empty union.

>When I do the following I don't statically get an error on the assert_never(my_var) call.

Are you sure about that? When I copy your example, I receive an error.

<img width="811" alt="Screen Shot 2021-02-26 at 1 30 30 AM" src="https://user-images.githubusercontent.com/7040122/109275637-3c186000-77d2-11eb-8b93-918ae5710830.png">

**Comment by bobbyrenwick** [comment]
Ah yes, I'd set my config to the following to calm down the number of issues I was seeing:

```
"typeCheckingMode": "basic",
"reportGeneralTypeIssues": false
```

Thanks for the explanations and help!

**Comment by erictraut** [comment]
I consulted with my colleagues in the TypeScript team, and they gave me some ideas for how to tackle this. While it's true that TypeScript doesn't handle this case for chains of if/else statements, it does handle it for switch statements. I should be able to borrow their approach. I will probably need to limit it to if/else chains that involve the same expression pattern multiple times (like in your example above), but I think that's a reasonable limitation.

**Comment by bobbyrenwick** [comment]
That sounds great. Thanks @erictraut! I think the limitation on the if/else chain doing the same expression pattern makes sense. Will the TS team also be looking to do the same thing, I wonder?

**Comment by erictraut** [comment]
The TS team said that most TS programs use switch statements in this case, so they haven't received much of a push from TS developers to add support for chained if/else. Python doesn't have a switch statement, so it's understandable why if/else chains are much more common.

**Comment by erictraut** [comment]
With the help of the TS team, I figured out how to implement this feature. It will be included in the next release.

See here for documentation: https://github.com/microsoft/pyright/blob/master/docs/type-concepts.md#narrowing-for-implied-else

**Comment by bobbyrenwick** [comment]
@erictraut that's brilliant news, thank you! Those docs are exactly how I'd imagine the narrowing to work 💥

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.118, which I just published. It will also be included in the next release of Pylance.
